lib/sourceror.ex

```diff
@@ -719,153 +745,146 @@ defmodule Sourceror do
+    to_patch = IO.iodata_to_binary([first_line, middle_lines, last_line])
 
     patch_text =
       if is_binary(patch.change) do
         patch.change
       else
-        original_text = Enum.join([first_to_patch | to_patch], "\n")
-        patch.change.(original_text)
+        patch.change.(to_patch)
       end
 
-    [first_patch | middle_patch] = String.split(patch_text, ~r/\n|\r\n|\r/)
-
-    middle_patch =
-      if is_binary(patch.change) and Map.get(patch, :preserve_indentation, true) do
-        indent = get_indent(first)
-
-        indent =
-          if String.trim(first) != "" and get_indent(List.first(middle_patch) || "") > 0 do
-            # If the patch does not start at the start of the line and the next
-            # lines have an additional indentation, then we need to add it to
-            # prevent the "flattening" of the indentations, essentially to
-            # avoid this:
-            #     foo do bar do
-            #       :ok
-            #     end
-            #     end
-            indent + 1
-          else
-            indent
-          end
+    patched =
+      case split_lines(patch_text) do
+        [patched] ->
+          patched
 
-        Enum.map_join(middle_patch, "\n", &(String.duplicate("\s\s", indent) <> &1))
-      else
-        middle_patch
-        |> Enum.join("\n")
+        [first_patched | rest_patched] ->
+          rest_patched =
+            if is_binary(patch.change) && Map.get(patch, :preserve_indentation, true) do
+              indent = get_indent(prefix)
+              Enum.map(rest_patched, &[indentation(indent), &1])
+            else
+              rest_patched
+            end
+
+          [first_patched | rest_patched]
       end
 
-    [first <> first_patch, middle_patch <> last | rest]
+    [prefix, patched, suffix, rest]
+    |> IO.iodata_to_binary()
+    |> split_lines()
+  end
+
+  defp relevant_patch_lines(%{range: range}, lines) do
+    case range.end[:line] - range.start[:line] + 1 do
+      1 ->
+        [line | rest] = lines
+        {prefix, rest_line} = String.split_at(line, range.start[:column] - 1)
+
+        {to_patch, suffix} =
+          String.split_at(rest_line, range.end[:column] - String.length(prefix) - 1)
+
+        {{prefix, to_patch}, [], {"", suffix}, rest}
+
+      line_span ->
+        {[first_line | rest_to_patch], rest} = Enum.split(lines, line_span)
+        {last_line, rest_to_patch} = List.pop_at(rest_to_patch, -1, "")
+        {prefix, first_to_patch} = String.split_at(first_line, range.start[:column] - 1)
+        {last_to_patch, suffix} = String.split_at(last_line, range.end[:column] - 1)
+
+        {{prefix, first_to_patch}, rest_to_patch, {last_to_patch, suffix}, rest}
+    end
   end
 
   defp get_indent(string, count \\ 0)
   defp get_indent("\s\s" <> rest, count), do: get_indent(rest, count + 1)
   defp get_indent(_, count), do: count
+
+  defp indentation(indent), do: String.duplicate("\s\s", indent)
+
+  defp split_lines(string) do
```

Comment:
This function splits on newlines but retains those newlines at the end of their respective line. This allows for some simpler shared logic between single and multi-line patches, removing a special case where an additional newline would have to be inserted in certain multiline situations.